certora/specs/Tokens.spec

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+import "ConsistentState.spec";
+
+using UtilHarness as Util;
+
+methods {
+    function Util.balanceOf(address, address) external returns(uint256) envfree;
+    function Util.safeTransferFrom(address, address, address, uint256) external envfree;
+    function Util.withdrawnAssets(address, MetaMorphoHarness.Id, uint256, uint256) external returns (uint256) envfree;
+
+    function _.transfer(address, uint256) external => DISPATCHER(true);
+    function _.transferFrom(address, address, uint256) external => DISPATCHER(true);
+    function _.balanceOf(address) external => DISPATCHER(true);
+
+    function _.supply(MetaMorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, bytes data) external => summarySupply(marketParams, assets, shares, onBehalf, data) expect (uint256, uint256) ALL;
+    function _.withdraw(MetaMorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, address receiver) external => summaryWithdraw(marketParams, assets, shares, onBehalf, receiver) expect (uint256, uint256) ALL;
+    function _.idToMarketParams(MetaMorphoHarness.Id id) external => summaryIdToMarketParams(id) expect MetaMorphoHarness.MarketParams ALL;
+
+    function _.expectedSupplyAssets(MetaMorphoHarness.MarketParams, address) external => NONDET;
+    function _.borrowRate(MetaMorphoHarness.MarketParams, MetaMorphoHarness.Market) external => NONDET;
+}
+
+function summaryIdToMarketParams(MetaMorphoHarness.Id id) returns MetaMorphoHarness.MarketParams {
+    MetaMorphoHarness.MarketParams marketParams;
+    uint256 lastUpdated = Morpho.lastUpdate(id);
+
+    // Safe require because markets in the supply/withdraw queue have positive last update (see LastUpdated.spec).
+    require lastUpdated > 0;
+    // Safe require because it is a verified invariant in Morpho Blue.
+    require lastUpdated > 0 => Morpho.libId(marketParams) == id;
+
+    return marketParams;
+}
+
+function summarySupply(MetaMorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, bytes data) returns (uint256, uint256) {
+    assert shares == 0;
+    assert onBehalf == currentContract;
+    assert data.length == 0;
+
+    // Safe require because it is a verified invariant.
+    require hasSupplyCapIsEnabled(Morpho.libId(marketParams));
```

Comment:
I don't understand completely the sense of a "require" in a summary. It makes total sense for me for an invariant or a rule, we assume more things to verify or prove something, but in a summary where we are describing an execution I don't really get the sense.

What happens for example if you require a non-sense.

Reply:
`require` from summary behaves the same as a regular require in the rule, except that it is as if it was written inside the function call. Usually it's useful to have access to the proper parameters, because inside the function you can have more context.

As for require elsewhere, if you `require false` then you can prove anything you want, which is obviously dangerous

Reply:
I get it